python/cusignal/windows/windows.py

```diff
@@ -53,6 +41,21 @@ def _truncate(w, needed):
         return w
 
 
+_general_cosine_kernel = cp.ElementwiseKernel(
+    "float64 delta, float64 start, raw T a, int32 N",
+    "T w",
+    """
+    double fac = start + delta * i;
+    T temp = 0.0;
+    for (int k = 0; k < N; k++) {
+        temp += a[k] * cos(k * fac);
+    }
+    w = temp;
+    """,
+    "_general_cosine_kernel",
+)
+
+
```

Comment:
Question: How does this handle single or half precision data? Is the dtype casted up to double for the elementwise kernel?

Reply:
@awthomp I've updated all the ElementwiseKernels to take generic input types.

But based on the each function's logic it will not change anything. All output arrays are preset to _float64_.
Do you want to get the user an option on data type? It would just require adding an optional input parameter to each function.

Also, I'm not sure NumPy or CuPy support half precision as a native type. Note, it wouldn't change the compute of the kernel unless CuPy packs two _half_ into a _half2_ per thread, which I'm pretty confident they are not :/